File: Web_Stadium/Services/KnockOutService.cs
using Microsoft.EntityFrameworkCore;
using Web_Stadium.EFCore;

namespace Web_Stadium.Services
{
    /// <summary>
    /// Sinh lịch vòng Knock-out (Tứ kết → Bán kết → Chung kết)
    /// dựa trên kết quả vòng bảng — lấy nhất/nhì mỗi bảng
    /// </summary>
    public class KnockOutService
    {
        private readonly SanBongContext _context;
        private readonly StandingService _standingService;

        public KnockOutService(SanBongContext context, StandingService standingService)
        {
            _context = context;
            _standingService = standingService;
        }

        // ══════════════════════════════════════════════════════════
        // Sinh toàn bộ vòng knock-out từ kết quả vòng bảng
        // ══════════════════════════════════════════════════════════
        public async Task<(bool ok, string error)> SinhVongKnockOut(int giaiDauId)
        {
            var giai = await _context.GiaiDaus
                .Include(g => g.BangDaus)
                .Include(g => g.DoiBongs)
                .Include(g => g.TranDaus)
                .FirstOrDefaultAsync(g => g.Id == giaiDauId);

            if (giai == null) return (false, "Không tìm thấy giải!");

            // Kiểm tra đã có vòng knock-out chưa
            var daCoKO = giai.TranDaus.Any(t => t.LoaiVong != "VongBang");
            if (daCoKO) return (false, "Vòng knock-out đã được sinh rồi!");

            // Lấy BXH từng bảng → lấy nhất/nhì
            var bxh = await _standingService.GetStandings(giaiDauId);
            var doiVao = new List<(DoiBong doi, int thuHang, string bang)>();

            foreach (var bang in giai.BangDaus)
            {
                if (!bxh.TryGetValue(bang.Id, out var rows) || rows.Count < 2)
                    return (false, $"{bang.TenBang} chưa đủ kết quả!");

                // Nhất bảng
                var nhat = giai.DoiBongs.FirstOrDefault(d => d.Id == rows[0].DoiId);
                var nhi = giai.DoiBongs.FirstOrDefault(d => d.Id == rows[1].DoiId);
                if (nhat == null || nhi == null)
                    return (false, $"Không xác định được đội từ {bang.TenBang}!");

                doiVao.Add((nhat, 1, bang.TenBang));
                doiVao.Add((nhi, 2, bang.TenBang));
            }

            // Xác định format dựa số đội vào knock-out
            int soDoiKO = doiVao.Count;
            var tranKO = new List<TranDau>();
            var ngayKO = giai.NgayKetThuc.AddDays(-21); // 3 tuần trước khi kết thúc

            if (soDoiKO == 8)
            {
                // 4 bảng × 2 = 8 đội → Tứ kết (4 trận) → Bán kết (2 trận) → CK (1 trận)
                tranKO.AddRange(SinhTuKet8Doi(doiVao, giaiDauId, ngayKO));
                tranKO.AddRange(SinhBanKetPlaceholder(giaiDauId, ngayKO.AddDays(7)));
                tranKO.Add(SinhChungKetPlaceholder(giaiDauId, ngayKO.AddDays(14)));
            }
            else if (soDoiKO == 4)
            {
                // 2 bảng × 2 = 4 đội → Bán kết (2 trận) → CK (1 trận)
                tranKO.AddRange(SinhBanKet4Doi(doiVao, giaiDauId, ngayKO));
                tranKO.Add(SinhChungKetPlaceholder(giaiDauId, ngayKO.AddDays(7)));
            }
            else if (soDoiKO == 2)
            {
                // 1 bảng chơi vòng tròn → 2 đội → Chung kết
                tranKO.Add(new TranDau
                {
                    GiaiDauId = giaiDauId,
                    DoiNhaId = doiVao[0].doi.Id,
                    DoiKhachId = doiVao[1].doi.Id,
                    VongDau = 99,
                    LoaiVong = "ChungKet",
                    NgayThiDau = ngayKO,
                    TrangThai = "Scheduled"
                });
            }
            else
            {
                return (false, $"Số đội vào knock-out ({soDoiKO}) không hợp lệ! Cần 2, 4 hoặc 8 đội.");
            }

            _context.TranDaus.AddRange(tranKO);
            await _context.SaveChangesAsync();

            return (true, "");
        }

        // ── 8 đội → 4 trận Tứ kết ─────────────────────────────────
        // Seeding: A1 vs D2, B1 vs C2, C1 vs B2, D1 vs A2
        private List<TranDau> SinhTuKet8Doi(
            List<(DoiBong doi, int thuHang, string bang)> doiVao,
            int giaiDauId, DateTime ngay)
        {
            // Lấy nhất/nhì từng bảng (A,B,C,D)
            var bangMap = doiVao.GroupBy(d => d.bang)
                .OrderBy(g => g.Key)
                .ToList();

            var result = new List<TranDau>();
            int n = bangMap.Count; // 4 bảng

            for (int i = 0; i < n / 2; i++)
            {
                var bangNhat = bangMap[i];
                var bangNhi = bangMap[n - 1 - i]; // cross-bracket

                var nhat1 = bangNhat.First(d => d.thuHang == 1).doi;
                var nhi2 = bangNhi.First(d => d.thuHang == 2).doi;
                var nhat2 = bangNhi.First(d => d.thuHang == 1).doi;
                var nhi1 = bangNhat.First(d => d.thuHang == 2).doi;

                result.Add(new TranDau
                {
                    GiaiDauId = giaiDauId,
                    DoiNhaId = nhat1.Id,
                    DoiKhachId = nhi2.Id,
                    VongDau = 10, // Tứ kết
                    LoaiVong = "TuKet",
                    NgayThiDau = ngay,
                    TrangThai = "Scheduled"
                });
                result.Add(new TranDau
                {
                    GiaiDauId = giaiDauId,
                    DoiNhaId = nhat2.Id,
                    DoiKhachId = nhi1.Id,
                    VongDau = 10,
                    LoaiVong = "TuKet",
                    NgayThiDau = ngay,
                    TrangThai = "Scheduled"
                });
            }

            return result;
        }

        // ── 4 đội → 2 trận Bán kết ───────────────────────────────
        private List<TranDau> SinhBanKet4Doi(
            List<(DoiBong doi, int thuHang, string bang)> doiVao,
            int giaiDauId, DateTime ngay)
        {
            var bangMap = doiVao.GroupBy(d => d.bang)
                .OrderBy(g => g.Key)
                .ToList();

            var nhat1 = bangMap[0].First(d => d.thuHang == 1).doi;
            var nhi2 = bangMap[1].First(d => d.thuHang == 2).doi;
            var nhat2 = bangMap[1].First(d => d.thuHang == 1).doi;
            var nhi1 = bangMap[0].First(d => d.thuHang == 2).doi;

            return new List<TranDau> {
                new TranDau {
                    GiaiDauId  = giaiDauId,
                    DoiNhaId   = nhat1.Id,
                    DoiKhachId = nhi2.Id,
                    VongDau    = 20,
                    LoaiVong   = "BanKet",
                    NgayThiDau = ngay,
                    TrangThai  = "Scheduled"
                },
                new TranDau {
                    GiaiDauId  = giaiDauId,
                    DoiNhaId   = nhat2.Id,
                    DoiKhachId = nhi1.Id,
                    VongDau    = 20,
                    LoaiVong   = "BanKet",
                    NgayThiDau = ngay,
                    TrangThai  = "Scheduled"
                }
            };
        }

        // ── Placeholder Bán kết (chờ kết quả Tứ kết) ────────────
        // Owner sẽ cập nhật đội sau khi có kết quả Tứ kết
        private List<TranDau> SinhBanKetPlaceholder(int giaiDauId, DateTime ngay)
        {
            // DoiNhaId/DoiKhachId = null → TBD (điền khi vòng trước Closed)
            return new List<TranDau> {
                new TranDau {
                    GiaiDauId  = giaiDauId,
                    DoiNhaId   = null,
                    DoiKhachId = null,
                    VongDau    = 20,
                    LoaiVong   = "BanKet",
                    NgayThiDau = ngay,
                    TrangThai  = "Pending" // Chờ xác định đội
                },
                new TranDau {
                    GiaiDauId  = giaiDauId,
                    DoiNhaId   = null,
                    DoiKhachId = null,
                    VongDau    = 20,
                    LoaiVong   = "BanKet",
                    NgayThiDau = ngay,
                    TrangThai  = "Pending"
                }
            };
        }

        // ── Placeholder Chung kết ─────────────────────────────────
        private TranDau SinhChungKetPlaceholder(int giaiDauId, DateTime ngay)
        {
            return new TranDau
            {
                GiaiDauId = giaiDauId,
                DoiNhaId = null,
                DoiKhachId = null,
                VongDau = 30,
                LoaiVong = "ChungKet",
                NgayThiDau = ngay,
                TrangThai = "Pending"
            };
        }

        // ══════════════════════════════════════════════════════════
        // Cập nhật đội cho trận knock-out sau khi có kết quả.
        // Gọi sau mỗi lần TranDau.TrangThai → Closed.
        //
        // Bracket cố định theo vị trí (không phụ thuộc thứ tự close):
        //   Tứ kết #0,#1 → Bán kết #0 (Nha, Khach)
        //   Tứ kết #2,#3 → Bán kết #1 (Nha, Khach)
        //   Bán kết #0,#1 → Chung kết (Nha, Khach)
        // Vị trí #i = index trong list các trận cùng LoaiVong sắp theo Id.
        // ══════════════════════════════════════════════════════════
        public async Task CapNhatDoiKnockOut(int tranDauId)
        {
            var tran = await _context.TranDaus
                .Include(t => t.GiaiDau).ThenInclude(g => g.TranDaus)
                .FirstOrDefaultAsync(t => t.Id == tranDauId);

            if (tran == null || tran.TrangThai != "Closed") return;
            if (!tran.BanThangNha.HasValue || !tran.BanThangKhach.HasValue) return;
            if (!tran.DoiNhaId.HasValue || !tran.DoiKhachId.HasValue) return;

            var vongTiepTheo = tran.LoaiVong switch
            {
                "TuKet" => "BanKet",
                "BanKet" => "ChungKet",
                _ => null
            };
            if (vongTiepTheo == null) return;

            // Đội thắng (hoà không xảy ra ở KO — nếu có thì lấy đội nhà cho chắc)
            int doiThang = tran.BanThangNha > tran.BanThangKhach
                ? tran.DoiNhaId.Value
                : tran.DoiKhachId.Value;

            // Vị trí (index) của trận nguồn trong list cùng LoaiVong
            var sameRound = tran.GiaiDau.TranDaus
                .Where(t => t.LoaiVong == tran.LoaiVong)
                .OrderBy(t => t.Id)
                .ToList();
            int srcIdx = sameRound.FindIndex(t => t.Id == tran.Id);
            if (srcIdx < 0) return;

            // Trận đích ở vòng kế: srcIdx / 2. Slot Nha nếu srcIdx chẵn, Khach nếu lẻ.
            var nextRound = tran.GiaiDau.TranDaus
                .Where(t => t.LoaiVong == vongTiepTheo)
                .OrderBy(t => t.Id)
                .ToList();
            int dstIdx = srcIdx / 2;
            if (dstIdx >= nextRound.Count) return;

            var dst = nextRound[dstIdx];
            if (srcIdx % 2 == 0) dst.DoiNhaId = doiThang;
            else dst.DoiKhachId = doiThang;

            // Nếu cả 2 slot đã có đội → chuyển từ Pending sang Scheduled
            if (dst.DoiNhaId.HasValue && dst.DoiKhachId.HasValue && dst.TrangThai == "Pending")
                dst.TrangThai = "Scheduled";

            await _context.SaveChangesAsync();
        }
    }
}

File: tournament-service/src/main/java/com/pitchhub/tournament/service/KnockOutService.java
package com.pitchhub.tournament.service;

import com.pitchhub.tournament.dto.response.BracketResponse;
import com.pitchhub.tournament.dto.response.StandingRowResponse;
import com.pitchhub.tournament.dto.response.TranDauResponse;
import com.pitchhub.tournament.exception.TournamentException;
import com.pitchhub.tournament.model.BangDau;
import com.pitchhub.tournament.model.DoiBong;
import com.pitchhub.tournament.model.GiaiDau;
import com.pitchhub.tournament.model.TranDau;
import com.pitchhub.tournament.repository.GiaiDauRepository;
import com.pitchhub.tournament.repository.TranDauRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Sinh lich vong Knock-out (Tu ket -> Ban ket -> Chung ket) dua tren ket
 * qua vong bang — lay nhat/nhi moi bang.
 *
 * TINH NANG HOAN THIEN: ben C# service nay da viet xong nhung KHONG duoc
 * dang ky DI, KHONG duoc TournamentService/Controller nao goi toi (dead
 * code). O day noi day day du: TournamentController goi sinhVongKnockOut()
 * sau khi vong bang xong, va TournamentStaffController goi
 * capNhatDoiKnockOut() moi khi mot tran knock-out duoc chot ket qua.
 */
@Service
public class KnockOutService {

    private final GiaiDauRepository giaiDauRepository;
    private final TranDauRepository tranDauRepository;
    private final StandingService standingService;

    public KnockOutService(GiaiDauRepository giaiDauRepository,
                            TranDauRepository tranDauRepository,
                            StandingService standingService) {
        this.giaiDauRepository = giaiDauRepository;
        this.tranDauRepository = tranDauRepository;
        this.standingService = standingService;
    }

    @Transactional
    public void sinhVongKnockOut(Integer giaiDauId) {
        GiaiDau giai = giaiDauRepository.findDetailById(giaiDauId)
                .orElseThrow(() -> new TournamentException(
                        TournamentException.Code.GIAI_KHONG_TON_TAI, "Khong tim thay giai!"));

        boolean daCoKO = giai.getTranDaus().stream().anyMatch(t -> !"VongBang".equals(t.getLoaiVong()));
        if (daCoKO) {
            throw new TournamentException(TournamentException.Code.TRANG_THAI_KHONG_HOP_LE,
                    "Vong knock-out da duoc sinh roi!");
        }

        Map<Integer, List<StandingRowResponse>> bxh = standingService.getStandings(giaiDauId);
        List<DoiVao> doiVao = new ArrayList<>();

        for (BangDau bang : giai.getBangDaus()) {
            List<StandingRowResponse> rows = bxh.get(bang.getId());
            if (rows == null || rows.size() < 2) {
                throw new TournamentException(TournamentException.Code.GIA_KHONG_DU_DOI,
                        bang.getTenBang() + " chua du ket qua!");
            }

            DoiBong nhat = timDoi(giai.getDoiBongs(), rows.get(0).getDoiId());
            DoiBong nhi = timDoi(giai.getDoiBongs(), rows.get(1).getDoiId());
            if (nhat == null || nhi == null) {
                throw new TournamentException(TournamentException.Code.GIA_KHONG_DU_DOI,
                        "Khong xac dinh duoc doi tu " + bang.getTenBang() + "!");
            }

            doiVao.add(new DoiVao(nhat, 1, bang.getTenBang()));
            doiVao.add(new DoiVao(nhi, 2, bang.getTenBang()));
        }

        int soDoiKO = doiVao.size();
        List<TranDau> tranKO = new ArrayList<>();
        LocalDateTime ngayKO = giai.getNgayKetThuc().minusDays(21);

        if (soDoiKO == 8) {
            tranKO.addAll(sinhTuKet8Doi(doiVao, giaiDauId, ngayKO));
            tranKO.addAll(sinhBanKetPlaceholder(giaiDauId, ngayKO.plusDays(7)));
            tranKO.add(sinhChungKetPlaceholder(giaiDauId, ngayKO.plusDays(14)));
        } else if (soDoiKO == 4) {
            tranKO.addAll(sinhBanKet4Doi(doiVao, giaiDauId, ngayKO));
            tranKO.add(sinhChungKetPlaceholder(giaiDauId, ngayKO.plusDays(7)));
        } else if (soDoiKO == 2) {
            TranDau chungKet = new TranDau();
            chungKet.setGiaiDauId(giaiDauId);
            chungKet.setDoiNha(doiVao.get(0).doi);
            chungKet.setDoiKhach(doiVao.get(1).doi);
            chungKet.setVongDau(99);
            chungKet.setLoaiVong("ChungKet");
            chungKet.setNgayThiDau(ngayKO);
            chungKet.setTrangThai("Scheduled");
            tranKO.add(chungKet);
        } else {
            throw new TournamentException(TournamentException.Code.GIA_KHONG_DU_DOI,
                    "So doi vao knock-out (" + soDoiKO + ") khong hop le! Can 2, 4 hoac 8 doi.");
        }

        tranDauRepository.saveAll(tranKO);
    }

    /** Goi sau khi mot tran TrangThai -> Closed — dien doi thang vao tran ke tiep. */
    @Transactional
    public void capNhatDoiKnockOut(Integer tranDauId) {
        TranDau tran = tranDauRepository.findById(tranDauId).orElse(null);
        if (tran == null || !"Closed".equals(tran.getTrangThai())) return;
        if (tran.getBanThangNha() == null) return;

        Integer doiThang = tran.getBanThangNha() > tran.getBanThangKhach() ? tran.getDoiNhaId() : tran.getDoiKhachId();
        if (doiThang == null) return;

        String vongTiepTheo = switch (tran.getLoaiVong()) {
            case "TuKet" -> "BanKet";
            case "BanKet" -> "ChungKet";
            default -> null;
        };
        if (vongTiepTheo == null) return;

        List<TranDau> ungVien = tranDauRepository.findByGiaiDauIdAndLoaiVong(tran.getGiaiDauId(), vongTiepTheo)
                .stream()
                .filter(t -> "Pending".equals(t.getTrangThai()))
                .sorted(Comparator.comparing(TranDau::getId))
                .toList();

        if (ungVien.isEmpty()) return;
        TranDau tranPending = ungVien.get(0);

        DoiBong doi = new DoiBong();
        doi.setId(doiThang);

        // "0" la sentinel TBD (DB NOT NULL DEFAULT 0, xem @DynamicInsert tren
        // TranDau) — coi 0 giong null khi kiem tra slot con trong.
        if (tranPending.getDoiNhaId() == null || tranPending.getDoiNhaId() == 0) {
            tranPending.setDoiNha(doi);
        } else if (tranPending.getDoiKhachId() == null || tranPending.getDoiKhachId() == 0) {
            tranPending.setDoiKhach(doi);
            tranPending.setTrangThai("Scheduled");
        }
        tranDauRepository.save(tranPending);
    }

    public BracketResponse getBracket(Integer giaiDauId) {
        List<TranDau> all = tranDauRepository.findByGiaiDauId(giaiDauId).stream()
                .filter(t -> !"VongBang".equals(t.getLoaiVong()))
                .sorted(Comparator.comparing(TranDau::getVongDau))
                .toList();

        Map<String, List<TranDau>> byRound = all.stream()
                .collect(Collectors.groupingBy(TranDau::getLoaiVong, LinkedHashMap::new, Collectors.toList()));

        List<BracketResponse.Round> rounds = new ArrayList<>();
        for (var e : byRound.entrySet()) {
            rounds.add(new BracketResponse.Round(e.getKey(),
                    e.getValue().stream().map(TranDauResponse::from).toList()));
        }

        return new BracketResponse(giaiDauId, rounds);
    }

    // ── 8 doi -> 4 tran Tu ket. Seeding: A1 vs D2, B1 vs C2, C1 vs B2, D1 vs A2 ──
    private List<TranDau> sinhTuKet8Doi(List<DoiVao> doiVao, Integer giaiDauId, LocalDateTime ngay) {
        List<List<DoiVao>> bangMap = groupByBang(doiVao);
        List<TranDau> result = new ArrayList<>();
        int n = bangMap.size();

        for (int i = 0; i < n / 2; i++) {
            List<DoiVao> bangNhat = bangMap.get(i);
            List<DoiVao> bangNhi = bangMap.get(n - 1 - i);

            DoiBong nhat1 = timTheoHang(bangNhat, 1);
            DoiBong nhi2 = timTheoHang(bangNhi, 2);
            DoiBong nhat2 = timTheoHang(bangNhi, 1);
            DoiBong nhi1 = timTheoHang(bangNhat, 2);

            result.add(taoTran(giaiDauId, nhat1, nhi2, 10, "TuKet", ngay));
            result.add(taoTran(giaiDauId, nhat2, nhi1, 10, "TuKet", ngay));
        }
        return result;
    }

    // ── 4 doi -> 2 tran Ban ket ──
    private List<TranDau> sinhBanKet4Doi(List<DoiVao> doiVao, Integer giaiDauId, LocalDateTime ngay) {
        List<List<DoiVao>> bangMap = groupByBang(doiVao);

        DoiBong nhat1 = timTheoHang(bangMap.get(0), 1);
        DoiBong nhi2 = timTheoHang(bangMap.get(1), 2);
        DoiBong nhat2 = timTheoHang(bangMap.get(1), 1);
        DoiBong nhi1 = timTheoHang(bangMap.get(0), 2);

        return List.of(
                taoTran(giaiDauId, nhat1, nhi2, 20, "BanKet", ngay),
                taoTran(giaiDauId, nhat2, nhi1, 20, "BanKet", ngay)
        );
    }

    // ── Placeholder Ban ket (cho ket qua Tu ket) ──
    private List<TranDau> sinhBanKetPlaceholder(Integer giaiDauId, LocalDateTime ngay) {
        return List.of(
                taoTranPending(giaiDauId, 20, "BanKet", ngay),
                taoTranPending(giaiDauId, 20, "BanKet", ngay)
        );
    }

    private TranDau sinhChungKetPlaceholder(Integer giaiDauId, LocalDateTime ngay) {
        return taoTranPending(giaiDauId, 30, "ChungKet", ngay);
    }

    private TranDau taoTran(Integer giaiDauId, DoiBong doiNha, DoiBong doiKhach,
                             int vongDau, String loaiVong, LocalDateTime ngay) {
        TranDau t = new TranDau();
        t.setGiaiDauId(giaiDauId);
        t.setDoiNha(doiNha);
        t.setDoiKhach(doiKhach);
        t.setVongDau(vongDau);
        t.setLoaiVong(loaiVong);
        t.setNgayThiDau(ngay);
        t.setTrangThai("Scheduled");
        return t;
    }

    private TranDau taoTranPending(Integer giaiDauId, int vongDau, String loaiVong, LocalDateTime ngay) {
        TranDau t = new TranDau();
        t.setGiaiDauId(giaiDauId);
        t.setVongDau(vongDau);
        t.setLoaiVong(loaiVong);
        t.setNgayThiDau(ngay);
        t.setTrangThai("Pending");
        return t;
    }

    private List<List<DoiVao>> groupByBang(List<DoiVao> doiVao) {
        Map<String, List<DoiVao>> grouped = doiVao.stream()
                .collect(Collectors.groupingBy(DoiVao::bang, LinkedHashMap::new, Collectors.toList()));
        List<String> keys = new ArrayList<>(grouped.keySet());
        Collections.sort(keys);
        List<List<DoiVao>> result = new ArrayList<>();
        for (String k : keys) result.add(grouped.get(k));
        return result;
    }

    private DoiBong timTheoHang(List<DoiVao> nhom, int thuHang) {
        return nhom.stream().filter(d -> d.thuHang() == thuHang).findFirst()
                .map(DoiVao::doi).orElse(null);
    }

    private DoiBong timDoi(Collection<DoiBong> doiBongs, Integer id) {
        return doiBongs.stream().filter(d -> d.getId().equals(id)).findFirst().orElse(null);
    }

    private record DoiVao(DoiBong doi, int thuHang, String bang) {
    }
}
